File: patterns-php/ph53-iface-tail-uninit/controls/wild_choice.py
#!/usr/bin/env python3
"""ph53 control driver: build and drive `controls/wild_choice.c`, with the
negatives that make it a check rather than a printout.

    python3 patterns-php/ph53-iface-tail-uninit/controls/wild_choice.py
    python3 .../wild_choice.py --selftest --reps 5

Read `controls/wild_choice.c`'s header first: it CHOOSES the value the
unwritten interface slot holds, by priming PHP 5.0.0's own size-class cache
(`zend_alloc.c:150-168`, `:263-279`) with a known payload and letting
`zend_compile.c:2571`'s `erealloc` pick the block straight back up.
`PROTOCOL_PHP.md` §H and `TASK_PHP_041.md` §3.5: *a control needs a chosen
value.*

⭐⭐ **THE RESULT, AND IT IS SHARPER THAN "R1h GIVES A CORRECT NO-MATCH".**
With the value chosen rather than observed, `d09cdd9f71f3` is **COMPLETE on the
comparing consumer and USELESS on the dereferencing one** — and on the
dereferencing one it makes things *differently* bad rather than better: R1 with
a plausible wild pointer returns a **silent wrong answer**, and R1h **always
dies**. An integrity bug becomes a guaranteed availability bug.

The negatives:

    C1 MUST-FIRE     R1 primed with &pool[3], target pool[3]: the COMPARE-ONLY
                     consumer must MATCH AN UNWRITTEN SLOT (stop == 1). That is
                     a silent wrong answer from the consumer that does not
                     dereference, and `controls/r1h_consumers.py` cannot show
                     it because there the unprimed garbage happened not to match
    C2 MUST-FIRE     changing the priming to &pool[5] must CHANGE the answer.
                     Without this the value could still be observed rather than
                     chosen, and C1 would be an accident
    C3 MUST-NOT-FIRE R1h must answer CORRECTLY (stop == 2) under BOTH primings
                     -- the memset makes the block's history irrelevant, which
                     is the whole of what the fix buys
    C4 MUST-FIRE     R1h's dereferencing consumer must DIE with SIGSEGV under
                     both primings. Silence here would mean the fix removed the
                     fault, which is the claim this row exists to refute
    C5 MUST-NOT-FIRE R1's dereferencing consumer, given a chosen VALID pool
                     address, must NOT die -- a wild dereference that lands on
                     live memory is a wrong answer and not a crash, and that
                     asymmetry is why the row ships two consumers
    C6 MUST-NOT-FIRE every line must be byte-identical across `--reps` runs.
                     A "chosen" value that moves between runs was observed
"""

import argparse
import os
import re
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROW = os.path.dirname(HERE)
REPO = os.path.dirname(os.path.dirname(ROW))
OUT = os.path.join(REPO, ".temp", "php41", "controls")
CLANG = os.path.expanduser("~/tools/llvm/bin/clang")

ROW_RE = re.compile(
    r"primed with &pool\[(\d)\]\s+cmp_scan=(\d+) \((.*?)\)\s+"
    r"deref (?:hit=(\d) id=([0-9a-f]+)|DIED signal=(\d+))")


def build():
    os.makedirs(OUT, exist_ok=True)
    out = os.path.join(OUT, "wild_choice")
    cmd = [CLANG, "-std=c99", "-O1", "-Wall", "-Wextra",
           "-I", os.path.join(REPO, "common-php"),
           "-o", out, os.path.join(HERE, "wild_choice.c")]
    r = subprocess.run(cmd, capture_output=True, text=True)
    if r.returncode != 0:
        raise SystemExit(f"BUILD FAILED:\n{r.stdout}{r.stderr}")
    if r.stderr.strip():
        print("  (compiler diagnostics)")
        print("    " + r.stderr.strip().replace("\n", "\n    "))
    return out


def parse(text):
    """[(arm, chosen, stop, verdict, hit, id, signal)] in file order, where
    `arm` is 'R1' before the R1h header and 'R1h' after it."""
    rows, arm = [], "R1"
    for line in text.split("\n"):
        if line.startswith("R1h "):
            arm = "R1h"
            continue
        m = ROW_RE.search(line)
        if m:
            rows.append((arm, int(m.group(1)), int(m.group(2)), m.group(3),
                         None if m.group(4) is None else int(m.group(4)),
                         m.group(5),
                         None if m.group(6) is None else int(m.group(6))))
    return rows


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--selftest", action="store_true")
    ap.add_argument("--reps", type=int, default=5)
    a = ap.parse_args()

    binary = build()
    outs = []
    for _ in range(max(1, a.reps)):
        r = subprocess.run([binary], capture_output=True, text=True, timeout=300)
        if r.returncode != 0:
            raise SystemExit(f"probe exited {r.returncode}:\n{r.stdout}{r.stderr}")
        outs.append(r.stdout)
    print(outs[0], end="")

    if not a.selftest:
        return 0

    print("-- negatives (PROTOCOL_PHP.md H) " + "-" * 46)
    bad = []
    rows = parse(outs[0])
    if len(rows) != 4:
        raise SystemExit(f"expected 4 rows, parsed {len(rows)}: {rows}")
    by = {(arm, ch): r for r in rows for arm, ch in [(r[0], r[1])]}

    # C1
    r = by[("R1", 3)]
    if r[2] != 1:
        bad.append(f"C1: R1 primed with &pool[3] gave cmp_scan={r[2]}, want 1 "
                   f"(a MATCH on the unwritten slot). Without it the "
                   f"compare-only consumer looks harmless, and the row's claim "
                   f"that d09cdd9f71f3 buys something REAL there has no "
                   f"evidence")
    else:
        print("  ok   C1 MUST-FIRE: R1 matches an UNWRITTEN slot -- a silent "
              "wrong answer from the consumer that does not dereference")

    # C2
    if by[("R1", 3)][2] == by[("R1", 5)][2]:
        bad.append(f"C2: both primings gave cmp_scan={by[('R1', 3)][2]}, so the "
                   f"value is not demonstrably CHOSEN. The whole point of this "
                   f"probe over controls/r1h_consumers.py is that changing the "
                   f"priming changes the answer")
    else:
        print(f"  ok   C2 MUST-FIRE: the answer moves with the priming "
              f"({by[('R1', 3)][2]} vs {by[('R1', 5)][2]}) -- the value is "
              f"CHOSEN, not observed")

    # C3
    for ch in (3, 5):
        if by[("R1h", ch)][2] != 2:
            bad.append(f"C3: R1h primed with &pool[{ch}] gave "
                       f"cmp_scan={by[('R1h', ch)][2]}, want 2 (no match). The "
                       f"memset is supposed to make the block's history "
                       f"irrelevant")
    if not any(b.startswith("C3") for b in bad):
        print("  ok   C3 MUST-NOT-FIRE: R1h answers correctly under BOTH "
              "primings -- history is irrelevant once the block is zeroed")

    # C4
    for ch in (3, 5):
        if by[("R1h", ch)][6] != 11:
            bad.append(f"C4: R1h primed with &pool[{ch}]: the dereferencing "
                       f"consumer did NOT die with SIGSEGV (signal="
                       f"{by[('R1h', ch)][6]!r}, hit={by[('R1h', ch)][4]!r}). "
                       f"The row's headline is that the fix does NOT remove "
                       f"the fault, and this is where that is measured")
    if not any(b.startswith("C4") for b in bad):
        print("  ok   C4 MUST-FIRE: R1h's dereferencing consumer dies with "
              "SIGSEGV under both primings -- the NULL is deterministic and "
              "the fault survives the fix")

    # C5
    r = by[("R1", 3)]
    if r[6] is not None:
        bad.append(f"C5: R1's dereferencing consumer died (signal={r[6]}) on a "
                   f"CHOSEN VALID pool address. It should return a wrong "
                   f"ANSWER, not a crash -- that asymmetry between the two "
                   f"consumers is why the kernel ships both")
    elif r[4] != 1:
        bad.append(f"C5: R1's dereferencing consumer returned hit={r[4]}, want "
                   f"1 -- the chosen pointer names pool[3] and the target IS "
                   f"pool[3]")
    else:
        print(f"  ok   C5 MUST-NOT-FIRE: R1's wild dereference lands on live "
              f"memory and returns a WRONG ANSWER (id={r[5]}), not a crash")

    # C6
    if len(set(outs)) != 1:
        bad.append(f"C6: {len(set(outs))} distinct outputs over {a.reps} runs. "
                   f"A CHOSEN value does not move between runs; this one did, "
                   f"so the probe is observing something")
    else:
        print(f"  ok   C6 MUST-NOT-FIRE: byte-identical over {a.reps} runs")

    print()
    if bad:
        print("SELFTEST FAIL")
        for b in bad:
            print(f"    {b}")
        return 1
    print("SELFTEST PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: patterns-php/ph53-iface-tail-uninit/controls/wild_choice.c
/* ph53 control -- THE WILD POINTER, **CHOSEN RATHER THAN OBSERVED**.
 *
 *   ~/tools/llvm/bin/clang -std=c99 -O1 -I ../../../common-php \
 *       -o wild_choice controls/wild_choice.c && ./wild_choice
 *   (or through `controls/wild_choice.py`, which also runs the negatives)
 *
 * ============================================================================
 * ⚠⚠ WHY A CONTROL NEEDS A CHOSEN VALUE
 * ============================================================================
 * The defect is an uninitialised read, so what the slot holds is whatever the
 * allocator left there -- and a control that merely OBSERVES that is a control
 * whose answer changes with the libc, the heap history and the sanitizer.
 * `controls/r1h_consumers.py` shows the OBSERVED case (under ASan it is
 * deterministic at `0xbe` fill bytes, and under plain `malloc` it is not).
 * This probe CHOOSES it.
 *
 * The mechanism is PHP 5.0.0's own size-class cache, `zend_alloc.c:150-168`
 * and `:263-279`, modelled in `common-php/emalloc_shim.h`: `_efree` on a block
 * whose recorded size falls in a cache class does **not** return it to
 * `malloc` -- it pushes the block onto `AG(cache)[real_size>>3]` **with its
 * payload intact** -- and the next same-class `_emalloc` pops it straight back.
 * So:
 *
 *     1. emalloc(8 * n_decl)          same size class as the interface array
 *     2. write a CHOSEN value into every slot
 *     3. efree                        -> the shim's cache, contents intact
 *     4. erealloc(NULL, 8 * n_decl)   -> zend_compile.c:2571, cache HIT
 *     5. the "uninitialised" tail now holds exactly what step 2 chose
 *
 * ⚠ Step 4 is why `php_shim_reset()` is NOT called between 3 and 4: the kernel
 * calls it at the top of every call precisely so that call N does not depend on
 * call N-1, and this probe is deliberately the configuration that does.
 *
 * ============================================================================
 * ⭐⭐ WHAT IT SHOWS, AND IT IS SHARPER THAN "R1h GIVES A CORRECT NO-MATCH"
 * ============================================================================
 * With the value chosen, the COMPARE-ONLY consumer (`zend_compile.c:1951`,
 * which does NOT dereference) gives a **silent wrong answer** on R1: the
 * unwritten slot compares EQUAL to an interface the class never implemented,
 * so `zend_do_inherit_interfaces` reports a duplicate that is not there. On
 * R1h the slot is NULL, which can never equal a pool address, so the answer is
 * RIGHT.
 *
 * ▶ **So the 2005 fix is COMPLETE on the comparing consumer and USELESS on the
 * dereferencing one** -- which is a stronger statement than
 * `controls/r1h_consumers.py` can make, because there the unprimed garbage
 * happened not to match and "correct" could have been luck.
 *
 * The dereferencing consumer (`zend_operators.c:1535`) is reported for both
 * primings too: with a chosen VALID pool address it returns the chosen id and
 * does not crash -- a silent wrong answer from a wild dereference that happened
 * to land on live memory, which is what CWE-824 costs when it does not kill
 * the process.
 *
 * ⚠ This TU is a STANDALONE PROBE and is not on `harness/build.py`'s path --
 * that compiles exactly three translation units (`PROTOCOL_PHP.md` §B2). It is
 * under `controls/` and `harness-php/gate.py::c_digest_audit` does not reach
 * it, so nothing it does can move a measured number.
 */
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#define PHP_SHIM_IMPL
#include "emalloc_shim.h"

#define MAXP 8

typedef struct { uint64_t id; } iface;
typedef struct { iface **interfaces; unsigned int num_interfaces; } ce_t;

static iface pool[MAXP];

/* zend_operators.c:1534-1535, narrowed -- THE FAULTING CONSUMER. */
static int deref_scan(const ce_t *ce, const iface *target, uint64_t *got)
{
    unsigned i;
    for (i = 0; i < ce->num_interfaces; i++) {
        if (ce->interfaces[i]->id == target->id) {   /* the dereference */
            *got = ce->interfaces[i]->id;
            return 1;
        }
    }
    *got = 0;
    return 0;
}

/* zend_compile.c:1951, narrowed -- THE COMPARE-ONLY CONSUMER. */
static unsigned cmp_scan(const ce_t *ce, const iface *entry)
{
    unsigned i;
    for (i = 0; i < ce->num_interfaces; i++) {
        if (ce->interfaces[i] == entry) {            /* no dereference */
            break;
        }
    }
    return i;
}

/* One run.  `chosen` is the pool index every primed slot is set to point at,
 * or -1 for "do not prime".  `hardened` selects `d09cdd9f71f3`'s phase 2.
 * `fill0` writes slot 0 legitimately, leaving the rest uncovered.
 *
 * ⚠ THE DEREFERENCING CONSUMER IS RUN IN A FORKED CHILD, because on the
 * hardened arm it dereferences NULL and dies -- which is the row's result and
 * must be REPORTED rather than allowed to take the probe with it.  The parent
 * prints the child's wait status, so `signal=11` is a measurement. */
static void run(int chosen, int hardened, unsigned n_decl, int fill0,
                int target, const char *label)
{
    ce_t ce;
    unsigned i;
    unsigned stop;
    pid_t pid;
    int status = 0;
    size_t bytes = sizeof(iface *) * (size_t)n_decl;

    php_shim_reset();

    if (chosen >= 0) {
        /* steps 1-3: prime the size-class cache with a CHOSEN payload */
        iface **prime = (iface **) php_shim_emalloc(bytes);
        for (i = 0; i < n_decl; i++)
            prime[i] = &pool[chosen];
        php_shim_efree(prime);          /* -> the shim's cache, intact */
    }

    /* zend_compile.c:3747-3748 */
    ce.interfaces = NULL;
    ce.num_interfaces = 0;
    for (i = 0; i < n_decl; i++)
        ce.num_interfaces++;            /* :2591, and no slot is written */

    /* zend_compile.c:2569-2572, R1 or R1h */
    if (ce.num_interfaces > 0) {
        if (hardened) {
            ce.interfaces = (iface **) php_shim_emalloc(bytes);
            memset(ce.interfaces, 0, bytes);
        } else {
            ce.interfaces = (iface **) php_shim_erealloc(ce.interfaces, bytes);
        }
    }

    if (fill0 && n_decl > 0)
        ce.interfaces[0] = &pool[0];    /* one legitimate ADD_INTERFACE */

    /* the COMPARE-ONLY consumer: never fatal, so it runs in-process */
    stop = cmp_scan(&ce, &pool[target]);
    printf("  %-24s cmp_scan=%u %-26s", label, stop,
           stop == ce.num_interfaces ? "(no match, CORRECT)"
                                     : (stop == 0 ? "(matched slot 0)"
                                                  : "(MATCHED UNWRITTEN SLOT)"));
    fflush(stdout);

    /* the DEREFERENCING consumer, in a child */
    pid = fork();
    if (pid == 0) {
        uint64_t got = 0;
        int hit = deref_scan(&ce, &pool[target], &got);
        printf("deref hit=%d id=%016llx\n", hit, (unsigned long long) got);
        fflush(stdout);
        _exit(0);
    }
    waitpid(pid, &status, 0);
    if (WIFSIGNALED(status))
        printf("deref DIED signal=%d\n", WTERMSIG(status));
    else if (WEXITSTATUS(status) != 0)
        printf("deref exit=%d\n", WEXITSTATUS(status));

    if (ce.num_interfaces > 0 && ce.interfaces)
        php_shim_efree(ce.interfaces);
}

int main(void)
{
    unsigned j;
    for (j = 0; j < MAXP; j++)
        pool[j].id = 0x0C1A5500000001ULL + j;

    printf("ph53 controls/wild_choice.c -- the wild pointer, CHOSEN\n");
    printf("  pool[j].id = 0x0C1A5500000001 + j, j in 0..7; n_decl = 2; "
           "slot 0 filled with &pool[0]; query target = pool[3]\n\n");

    printf("R1  (erealloc, no memset) -- the unwritten slot is whatever the "
           "cache handed back:\n");
    run(3, 0, 2, 1, 3, "primed with &pool[3]");
    run(5, 0, 2, 1, 3, "primed with &pool[5]");
    printf("\nR1h (d09cdd9f71f3: emalloc + memset) -- history is irrelevant:\n");
    run(3, 1, 2, 1, 3, "primed with &pool[3]");
    run(5, 1, 2, 1, 3, "primed with &pool[5]");
    return 0;
}
